README.md
# Clean interface to calculate percentages in Go

[![Go version](https://img.shields.io/github/go-mod/go-version/rorecek/percentage?style=flat-square)](https://github.com/rorecek/percentage/blob/master/go.mod)
[![Go Report Card](https://goreportcard.com/badge/github.com/rorecek/percentage?style=flat-square)](https://goreportcard.com/report/github.com/rorecek/percentage)
[![Godoc](http://img.shields.io/badge/go-documentation-blue.svg?style=flat-square)](https://godoc.org/github.com/rorecek/percentage)
[![Build Status](https://travis-ci.com/rorecek/percentage.svg?branch=master)](https://travis-ci.com/rorecek/percentage)
[![codecov](https://codecov.io/gh/rorecek/percentage/branch/master/graph/badge.svg)](https://codecov.io/gh/rorecek/percentage)
[![Releases](https://img.shields.io/github/v/release/rorecek/percentage?sort=semver&style=flat-square)](https://github.com/rorecek/percentage/releases)
[![LICENSE](https://img.shields.io/github/license/rorecek/percentage?style=flat-square)](https://github.com/rorecek/percentage/blob/master/LICENSE)

A clean interface to calculate percentages in Go: changes between values, percentage increases and partial values.
Think of it as a very lightweight math library for the basic stuff. Aimed to make your code more readable and easier to understand.
Inspired by [mattiasgeniar/php-percentages](https://github.com/mattiasgeniar/php-percentages)

## Installation

You can install this module with ```go get``` command:

``` bash
go get -u github.com/rorecek/percentage
```

## Usage

``` go
import "github.com/rorecek/percentage"

// What's the percentage increase from 100 to 120?
percentage.DifferenceBetween(100, 120);         // 20

// What's the absolute percentage change from 100 to 80?
percentage.AbsoluteDifferenceBetween(100, 80);  // 20, not -20

// How much is 120 compared to 100?
percentage.Calculate(120, 100);                 // 120

// How much is 50 compared to 100?
percentage.Calculate(50, 100);                  // 50

// What is 20% of 200?
percentage.Of(20, 200);                         // 40
```

## Testing

``` bash
go test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email pavel@rorecek.com instead of using the issue tracker.

## Credits

- [Pavel Rorecek](https://github.com/rorecek)
- [Mattias Geniar](https://github.com/mattiasgeniar/php-percentages)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `percentage` will be documented in this file

## 1.0.0 - 202X-XX-XX

- initial release


percentage_test.go
package percentage

import (
	"fmt"
	"testing"
)

func TestDifferenceBetween(t *testing.T) {
	var tests = []struct {
		a    float64
		b    float64
		want float64
	}{
		{100, 120, 20},
		{100, 80, -20},
	}

	for _, tt := range tests {
		name := fmt.Sprintf("%v, %v", tt.a, tt.b)

		t.Run(name, func(t *testing.T) {
			got := DifferenceBetween(tt.a, tt.b)

			if got != tt.want {
				t.Errorf("got %v, want %v", got, tt.want)
			}
		})
	}
}

func TestAbsoluteDifferenceBetween(t *testing.T) {
	var tests = []struct {
		a    float64
		b    float64
		want float64
	}{
		{100, 120, 20},
		{100, 80, 20},
	}

	for _, tt := range tests {
		name := fmt.Sprintf("%v, %v", tt.a, tt.b)

		t.Run(name, func(t *testing.T) {
			got := AbsoluteDifferenceBetween(tt.a, tt.b)

			if got != tt.want {
				t.Errorf("got %v, want %v", got, tt.want)
			}
		})
	}
}

func TestCalculate(t *testing.T) {
	var tests = []struct {
		new    float64
		origin float64
		want   float64
	}{
		{160, 80, 200},
		{60, 80, 75},
	}

	for _, tt := range tests {
		name := fmt.Sprintf("%v, %v", tt.new, tt.origin)

		t.Run(name, func(t *testing.T) {
			got := Calculate(tt.new, tt.origin)

			if got != tt.want {
				t.Errorf("got %v, want %v", got, tt.want)
			}
		})
	}
}

func TestOf(t *testing.T) {
	var tests = []struct {
		percentage float64
		base       float64
		want       float64
	}{
		{20, 200, 40},
		{50, 200, 100},
		{-20, 200, -40},
	}

	for _, tt := range tests {
		name := fmt.Sprintf("%v, %v", tt.percentage, tt.base)

		t.Run(name, func(t *testing.T) {
			got := Of(tt.percentage, tt.base)

			if got != tt.want {
				t.Errorf("got %v, want %v", got, tt.want)
			}
		})
	}
}

func ExampleDifferenceBetween_positive() {
	fmt.Println(DifferenceBetween(200, 240))
	// Output: 20
}

func ExampleDifferenceBetween_negative() {
	fmt.Println(DifferenceBetween(200, 160))
	// Output: -20
}

func ExampleAbsoluteDifferenceBetween() {
	fmt.Println(AbsoluteDifferenceBetween(200, 160))
	// Output: 20
}

func ExampleCalculate() {
	fmt.Println(Calculate(160, 80))
	// Output: 200
}

func ExampleOf() {
	fmt.Println(Of(20, 200))
	// Output: 40
}


percentage.go
// Copyright 2011 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

/*
Package percentage - a clean interface to calculate percentages in Go (changes between values, percentage increases and partial values).
Think of it as a very lightweight math library for the basic stuff. Aimed to make your code more readable and easier to understand.

Usage

What's the percentage increase from 100 to 120?
 percentage.DifferenceBetween(100, 120);         // 20

What's the absolute percentage change from 100 to 80?
 percentage.AbsoluteDifferenceBetween(100, 80);  // 20, not -20

How much is 120 compared to 100?
 percentage.Calculate(120, 100);                 // 120%

How much is 50 compared to 100?
 percentage.Calculate(50, 100);                  // 50%

What is 20% of 200?
 percentage.Of(20, 200);                         // 40
*/
package percentage

import "math"

// DifferenceBetween returns the percentage increase or decrease from a to b
func DifferenceBetween(a, b float64) float64 {
	return (b - a) / a * 100
}

// AbsoluteDifferenceBetween returns the absolute percentage increase or decrease from a to b
func AbsoluteDifferenceBetween(a, b float64) float64 {
	return math.Abs(DifferenceBetween(a, b))
}

// Calculate returns how much is the new value of the origin value in percentages
func Calculate(new, origin float64) float64 {
	return new / origin * 100
}

// Of returns a percentage from the base number
func Of(percentage, base float64) float64 {
	return percentage * base / 100
}
